### PR TITLE
Switch to named pipes

### PR DESCRIPTION
**Pull request by tylerl0706**
## PR Summary

switch to using named pipes on Windows (and Unix domain sockets) instead of TCP.

It's shockingly great how simple this change was :) 

marking WIP because I want to test on a few platforms and want a CI build.

## PR Checklist

Note: Tick the boxes below that apply to this pull request by putting an `x` between the square brackets.
Please mark anything not applicable to this PR `NA`.

- [x] PR has a meaningful title
- [x] Summarized changes
- [ ] This PR is ready to merge and is not work in progress
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready

### CODE REVIEW

File: src/session.ts [hunk 1]
```diff
@@ -165,19 +166,23 @@ export class SessionManager implements Middleware {
...
+                `-HostVersion '${this.hostVersion}'` +
+                `-AdditionalModules @('PowerShellEditorServices.VSCode') ` +
+                `-BundledModulesPath '${this.bundledModulesPath}'` +
+                `-EnableConsoleRepl ` +
+                `-LanguageServicePipeName LanguageService_${id}.pipe ` +
+                `-DebugServicePipeName DebugService_${id}.pipe `;
 
             if (this.sessionSettings.developer.editorServicesWaitForDebugger) {
-                this.editorServicesArgs += "-WaitForDebugger ";
+                this.editorServicesArgs += `-WaitForDebugger `;
```
Comment (rkeithhill) [inline src/session.ts]: No need for string interpolation here.

File: src/session.ts [hunk 2]
```diff
@@ -165,19 +166,23 @@ export class SessionManager implements Middleware {
...
             }
             if (this.sessionSettings.developer.editorServicesLogLevel) {
-                this.editorServicesArgs += "-LogLevel '" + this.sessionSettings.developer.editorServicesLogLevel + "' ";
+                this.editorServicesArgs += `-LogLevel '${this.sessionSettings.developer.editorServicesLogLevel}' `;
```
Comment (rkeithhill) [inline src/session.ts]: +1

**Comment by tylerl0706** [comment]
Connecting logic all seems right but there's a crash that occurs when a script finishes debugging:

> 2018-05-21 11:44:02 [NORMAL] - Method "Stop" at line 167 of C:\projects\PowerShellEditorServices\src\PowerShellEditorServices.Protocol\Server\DebugAdapter.cs
>
>    Debug adapter is shutting down...
>
>2018-05-21 11:44:02 [NORMAL] - Method "OnDebugServiceClientConnect" at line 311 of C:\projects\PowerShellEditorServices\src\PowerShellEditorServices.Host\EditorServicesHost.cs
>
>    Previous debug session ended, restarting debug service listener...
>
>2018-05-21 11:44:02 [VERBOSE] - Method "Start" at line 45 of C:\projects\PowerShellEditorServices\src\PowerShellEditorServices.Protocol\MessageProtocol\Channel\NamedPipeServerListener.cs
>
>    Named pipe server failed to start due to exception:
>    
>    All pipe instances are busy.
>2018-05-21 11:44:02 [ERROR] - Method "OnListenTaskCompleted" at line 391 of C:\projects\PowerShellEditorServices\src\PowerShellEditorServices.Protocol\MessageProtocol\ProtocolEndpoint.cs
>
>    ProtocolEndpoint message loop terminated due to unhandled exception:
>    
>    System.AggregateException: One or more errors occurred. ---> System.IO.IOException: All pipe instances are busy.     .........
    
I'll have to dig in to how the debug adapter handles ending of debugging.

**Comment by tylerl0706** [comment]
Ok that crash is fixed in https://github.com/PowerShell/PowerShellEditorServices/pull/668

I'm removing the WIP. I've tested named pipes on:
* Windows 10 with 5.1 & 6+
* macOS
* Ubuntu
* CentOS

File: src/session.ts (hunk 2 above)
Comment (tylerl0706) [reply src/session.ts]: why don't we need it here?

Comment (rkeithhill) [reply src/session.ts]: If your question is about the string interpolation, then I'm for its use here, where there is a variable to interpolate inside the string.  :-)

File: src/session.ts [hunk 3]
```diff
@@ -165,19 +166,23 @@ export class SessionManager implements Middleware {
...
+            // Generate a random id for the named pipes in case they have multiple instances of PSES running
+            const id = crypto.randomBytes(10).toString("hex");
             this.editorServicesArgs =
-                "-HostName 'Visual Studio Code Host' " +
-                "-HostProfileId 'Microsoft.VSCode' " +
-                "-HostVersion '" + this.hostVersion + "' " +
-                "-AdditionalModules @('PowerShellEditorServices.VSCode') " +
-                "-BundledModulesPath '" + this.bundledModulesPath + "' " +
-                "-EnableConsoleRepl ";
+                `-HostName 'Visual Studio Code Host' ` +
```
Comment (rjmholt) [inline src/session.ts]: Thoughts on using ordinary strings where no interpolation is needed vs consistency? Not opinionated either way.

File: src/utils.ts [hunk 4]
```diff
@@ -26,15 +26,9 @@ export function getPipePath(pipeName: string) {
...
-        // For details on how this path is computed, see the corefx
-        // source for System.IO.Pipes.PipeStream:
-        // tslint:disable-next-line:max-line-length
-        // https://github.com/dotnet/corefx/blob/d0dc5fc099946adc1035b34a8b1f6042eddb0c75/src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs#L340
-        return path.resolve(
-            os.tmpdir(),
-            ".dotnet", "corefx", "pipe",
-            pipeName);
+        // Windows uses NamedPipes where non-Windows platforms use Unix Domain Sockets.
+        // This requires connecting to the pipe file in different locations on Windows vs non-Windows.
```
Comment (rjmholt) [inline src/utils.ts]: Technically the name difference is just .NET Core being strange, rather than the use of domain sockets, no?

Comment (tylerl0706) [reply src/utils.ts]: nah the location is different too. Named Pipes are in:

`\\.\pipe\foo`

domain sockets are in:

`/tmp/CoreFxPipe_foo`

Even the Named Pipes on Windows using .NET Standard instead are placed here:

`\\.\pipe\foo`

File: src/session.ts (hunk 3 above)
Comment (tylerl0706) [reply src/session.ts]: for here, I just kept it consistent with the rest of the string concat

**Pull request merged**